### ISSUE TITLE
Existing rulesets are ignored when binding a solution

### ISSUE THREAD

**Issue by duncanp-sonar** [SonarSource/sonarlint-visualstudio#863]
### Description

If an unbound project has a custom existing ruleset, it will be ignored when the project is bound.

### TODO
Decide on the preferred behaviour e.g.
* include the generated shared ruleset in the existing custom rules, or
* create a new project-level ruleset that includes the existing custom ruleset and the generated shared ruleset, or
* ignore the existing custom ruleset and reference the generated shared ruleset directly.
* do nothing - current behaviour is good enough most cases.

What if the custom ruleset file is used by multiple projects (not all of which may be bound)?

### Related information

* SonarLint for VS Version: to current (v4.8)
* Languages in your solution: C#, VB.Net

### Other tickets in the same area
* #817: [Connected Mode] Stop generating project-level ruleset

**Comment by jirkapok** [comment]
I am expecting there are two main use cases:
* Each project has its own rules file
* All projects in a solution share the same rules file

Which one do you think the development team prefers? From our team point of view the expected behavior is:
* If there is no file create default one for all projects in solution directory (or ask user to choose the location)
* If there is one not project level default nor solution level default, skip update, since this is considered custom file
* If there is solution level default, update only this one

**Comment by duncanp-sonar** [comment]
@jirkapok I think I would go for a simple solution:
- if the project doesn't currently reference a ruleset file then reference the generated solution-level ruleset directly
- if the project does contain a custom ruleset then make that custom ruleset import the generated solution-level ruleset (regardless of whether the custom ruleset is specific to a single project or shared).

The reasoning for this is that the main purpose of binding is to make the local configuration the same as the server configuration, so all projects should reference the generated ruleset either directly or indirectly. However, there are circumstances when custom rulesets are still useful e.g. to configure additional third-party NuGet analyzers that are used by a project.

**Comment by jirkapok** [comment]
Yes, but for backward compatibility, where users already may have aplenty local project rule files, you dont want to merge them. Maybe in case all of them are equal.

**Comment by duncanp-sonar** [comment]
I'm not suggesting merging them. If a project already specifies a _CodeAnalysisRuleset_, I'm proposing we would leave that as the ruleset, but amend it so that that ruleset includes the generated ruleset.
If multiple project reference the same custom ruleset then the custom ruleset would only be updated once.

**Comment by duncanp-sonar** [comment]
This doesn't repro. On initial binding:
* if a project references a ruleset that is in or below the same folder as the project file (i.e. the ruleset is local to that project) then the ruleset will be updated to include the generated ruleset
* if a project references a ruleset that is outside the project folder then a new project-specific ruleset file will be created (since the existing ruleset is not project-specific there is a possibility that it is used in other solutions, so updating it might impact other solutions).
* if a ruleset file with the expected project-specific file name already exists on disc but isn't referenced by the `CodeAnalysisRuleSet` property in the project then a new project-level ruleset file will be created with a numeric suffix on the name to avoid clashing with the existing file e.g. `MyProject.ruleset` file already exists, generated ruleset will be called `MyProject-1.ruleset`

**Comment by rita-g-sonarsource** [comment]
Managed to reproduce: a new project-level ruleset is created when there is at least one rule-set declaration that does not reference the solution-level ruleset.

Steps to reproduce:
1. Create a project with a project-level ruleset.
2. Set one configuration to point to this ruleset. E.g.

```
	<PropertyGroup>
		<CodeAnalysisRuleSet>myruleset-project.ruleset</CodeAnalysisRuleSet>
	</PropertyGroup>
```
3. Create another configuration that points to solution level ruleset. E.g.
```
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|AnyCPU'">
    <CodeAnalysisRuleSet>..\.sonarlint\rita-g-sonarsource_sanity-connectedcsharp.ruleset</CodeAnalysisRuleSet>
  </PropertyGroup>
```
4. Bind the project

Expected: 
1. First ruleset is amended to include solution-level ruleset.
2. Second ruleset is untouched.

Actual:
1. First ruleset is amended
2. A new project-level ruleset is created that points to solution-level ruleset.
